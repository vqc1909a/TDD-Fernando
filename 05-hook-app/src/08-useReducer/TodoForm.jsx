
import React  from "react";
import PropTypes from "prop-types";
import { useForm } from "../hooks";
import uuid4 from "uuid4";

export const TodoForm = React.memo(({handleAddTodo}) => {
  const {description, handleInputChange, handleResetForm} = useForm({
    description: ''
  });

  const handleSubmit = (e) => {
    e.preventDefault();
    if(description.trim().length <= 0) return;

    const newTodo = {
      id: uuid4(),
      description,
      done: false
    }
    handleAddTodo(newTodo)
    handleResetForm();
  }

 
  return (
     <form onSubmit={handleSubmit}>
        <input 
            type="text" 
            placeholder="¿Qué hay que hacer?"
            className="form-control"
            name="description"
            value={description}
            onChange={handleInputChange}
        />
        <button 
            type="submit"
            className="btn btn-outline-primary mt-1"
        >
            Agregar
        </button>
    </form>
    )
})


TodoForm.propTypes = {
  handleAddTodo: PropTypes.func.isRequired
}