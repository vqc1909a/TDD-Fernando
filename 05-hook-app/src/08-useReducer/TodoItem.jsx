
import PropTypes from "prop-types";
import React from "react";

export const TodoItem = React.memo(({description, id, done ,handleDeleteTodo, handleToggleTodo}) => {
  return (
    <li aria-label="li" className="list-group-item d-flex justify-content-between">
      <span aria-label="span" style={{cursor: 'pointer'}} className={`align-self-center flex-fill ${done ? 'text-decoration-line-through': ''}`} onClick={() => handleToggleTodo(id)}>{description}</span>
      <button aria-label="button" className="btn btn-danger" onClick={() => handleDeleteTodo(id)}>Borrar</button>
    </li>
    )
})


TodoItem.propTypes = {
  description: PropTypes.string,
  id: PropTypes.string,
  done: PropTypes.bool,
  handleDeleteTodo: PropTypes.func.isRequired,
  handleToggleTodo: PropTypes.func.isRequired,
}

