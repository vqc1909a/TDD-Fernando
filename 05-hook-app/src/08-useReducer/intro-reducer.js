const initialState = [{
    id: 1,
    todo: 'Recolectar la piedra del alma',
    donce: false,
}]

const todoReducer = (state = initialState, action = {}) => {
    if(action.type === "add"){
        return [...state, action.payload]
    }
    //The reducer always return a new state object, it mean return a new object,  new array, etc whithout matter its content in these previous cases
    return state
}


let todos = todoReducer()
const newTodo = {
    id: 2,
    todo: 'Recolectar la piedra del poder',
    done: false
}

const addTodoAction = {
    type: 'add',
    payload: newTodo
}

todos = todoReducer(todos, addTodoAction);

console.log({
    todos
})