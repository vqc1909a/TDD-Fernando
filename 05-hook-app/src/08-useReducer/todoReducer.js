
export const todoReducer = (initialState = [], action) => {
  switch (action.type) {
    case 'addTodo':
        return [
            ...initialState,
            action.payload
        ]   
    case 'deleteTodo':
        return initialState.filter(todo => todo.id !== action.payload)

    case 'toggleTodo':
        return initialState.map(todo => todo.id.toString() === action.payload.toString() ? ({...todo, done: !todo.done}) : todo)

    default:
        return initialState
  }
    
}
