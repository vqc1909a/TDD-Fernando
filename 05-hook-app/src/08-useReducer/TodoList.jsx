
import PropTypes from "prop-types";
import { TodoItem } from "./TodoItem";
import React from "react";

export const TodoList = React.memo(({todos, handleDeleteTodo, handleToggleTodo}) => {
  // console.log("Ejecutando todolist")
  return (
    <ul className="list-group">
        {todos.map(todo => (
            <TodoItem key={todo.id} {...todo} handleDeleteTodo={handleDeleteTodo} handleToggleTodo={handleToggleTodo}/>
        ))}
    </ul>
    )
})


TodoList.propTypes = {
  todos: PropTypes.array.isRequired,
  handleDeleteTodo: PropTypes.func.isRequired,
  handleToggleTodo: PropTypes.func.isRequired
}