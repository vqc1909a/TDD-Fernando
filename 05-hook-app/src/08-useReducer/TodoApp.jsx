import {TodoList} from "./TodoList";
import { TodoForm } from "./TodoForm";
import { useTodos } from "../hooks";



export const TodoApp = () => {
    //It's recommended that we use the hook useTodos only in the father component, and then we pass the functions and the state to the children components, because we use the hook in the children components, will have two or more instances of own hook and it will trigger a leak memory
    const { handleAddTodo, handleDeleteTodo, handleToggleTodo, todos, todosCount, pendingTodosCount } = useTodos();
  
  return (
    <>
     <h1>TodoApp ({todosCount}) <small>pendientes: {pendingTodosCount}</small></h1>
     <hr />

     <div className="row">
        <div className="col-7">
            <TodoList todos={todos} handleDeleteTodo={handleDeleteTodo} handleToggleTodo={handleToggleTodo}/>
        </div>
        <div className="col-5">
            <h4>Agregar TODO</h4>
            <hr />
            <TodoForm handleAddTodo={handleAddTodo} />
        </div>
     </div>

    </>
  )
}
