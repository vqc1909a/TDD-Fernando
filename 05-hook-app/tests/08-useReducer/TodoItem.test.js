import { fireEvent, render, screen } from "@testing-library/react";
import { TodoItem } from "../../src/08-useReducer/TodoItem"

describe('tests on TodoItem component', () => { 
    let handleDeleteTodo = jest.fn();
    let handleToggleTodo = jest.fn();
    let id = "1";
    let description = "task1";
    let done = false;

    beforeEach(() => {
        jest.clearAllMocks();
    })

    test('should show the component by default', () => {
        const { container } = render(<TodoItem handleDeleteTodo={handleDeleteTodo} handleToggleTodo={handleToggleTodo} id={id} description={description} done={done} />);
        //We can get to the node element with the prop aria-label directly with the method getByLabelText
        const spanElement = screen.getByLabelText("span");
        expect(spanElement).toBeTruthy();
        expect(spanElement.textContent).toContain(description);
        expect(spanElement.classList.contains('text-decoration-line-through')).toBeFalsy();
        expect(spanElement.className).not.toContain('text-decoration-line-through');
        expect(container).toMatchSnapshot()
    })


    test('should show the component with todo completed', () => {
        done = true;
        render(<TodoItem handleDeleteTodo={handleDeleteTodo} handleToggleTodo={handleToggleTodo} id={id} description={description} done={done} />);
        //We can get to the node element with the prop aria-label directly with the method getByLabelText
        const spanElement = screen.getByLabelText("span");
        expect(spanElement).toBeTruthy();
        expect(spanElement.classList.contains('text-decoration-line-through')).toBeTruthy();
        expect(spanElement.className).toContain('text-decoration-line-through');
    })

    test('should call the function handleToggleTodo when trigger the click span', () => { 
        render(<TodoItem handleDeleteTodo={handleDeleteTodo} handleToggleTodo={handleToggleTodo} id={id} description={description} done={done} />);
        const spanElement = screen.getByLabelText("span");
        // spanElement.click(); //This property od node element is equal to fireEvent.click
        fireEvent.click(spanElement);
        expect(handleToggleTodo).toHaveBeenCalled();
        expect(handleToggleTodo).toHaveBeenCalledTimes(1);
        expect(handleToggleTodo).toHaveBeenCalledWith(id);
    })

    test('should call the function handleDeleteTodo when trigger the click button', () => { 
        render(<TodoItem handleDeleteTodo={handleDeleteTodo} handleToggleTodo={handleToggleTodo} id={id} description={description} done={done} />);
        const buttonElement = screen.getByLabelText("button");
        // buttonElement.click(); //This property od node element is equal to fireEvent.click
        fireEvent.click(buttonElement);
        expect(handleDeleteTodo).toHaveBeenCalled();
        expect(handleDeleteTodo).toHaveBeenCalledTimes(1);
        expect(handleDeleteTodo).toHaveBeenCalledWith(id);
    })
})