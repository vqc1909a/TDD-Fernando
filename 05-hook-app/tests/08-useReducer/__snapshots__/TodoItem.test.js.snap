// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on TodoItem component should show the component by default 1`] = `
<div>
  <li
    aria-label="li"
    class="list-group-item d-flex justify-content-between"
  >
    <span
      aria-label="span"
      class="align-self-center flex-fill "
      style="cursor: pointer;"
    >
      task1
    </span>
    <button
      aria-label="button"
      class="btn btn-danger"
    >
      Borrar
    </button>
  </li>
</div>
`;
